overlord/snapshotstate/snapshotstate.go

```diff
@@ -0,0 +1,264 @@
+
+	return nil
+}
+
+// List valid snapshots
+//
+// The snapshots are closed before returning.
+var List = backend.List
+
+// Save creates a taskset for taking snapshots of snaps' data.
+// Note that the state must be locked by the caller.
+func Save(st *state.State, snapNames []string, users []string) (uint64, []string, *state.TaskSet, error) {
+	var err error
+
+	if len(snapNames) == 0 {
+		snapNames, err = allActiveSnapNames(st)
+		if err != nil {
+			return 0, nil, nil, err
+		}
+	}
+
+	if err := snapstate.CheckChangeConflictMany(st, snapNames, nil); err != nil {
+		return 0, nil, nil, err
+	}
+
+	shID, err := newSnapshotID(st)
+	if err != nil {
+		return 0, nil, nil, err
+	}
+
+	ts := state.NewTaskSet()
+
+	for _, name := range snapNames {
+		desc := fmt.Sprintf("Saving the data of snap %q in snapshot #%d", name, shID)
+		task := st.NewTask("save-snapshot", desc)
+		shot := shotState{
+			ID:    shID,
+			Snap:  name,
+			Users: users,
+		}
+		task.Set("shot", &shot)
+		task.Set("snap-setup", &snapstate.SnapSetup{SideInfo: &snap.SideInfo{RealName: name}})
+		ts.AddTask(task)
+	}
+
+	return shID, snapNames, ts, nil
+}
+
+// Restore creates a taskset for restoring a snapshot's data.
+// Note that the state must be locked by the caller.
+func Restore(st *state.State, snapshotID uint64, snapNames []string, users []string) ([]string, *state.TaskSet, error) {
+	snapNames, filenames, err := snapNamesInSnapshot(snapshotID, snapNames)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if err := snapstate.CheckChangeConflictMany(st, snapNames, nil); err != nil {
+		return nil, nil, err
+	}
+
+	// restore needs to conflict with forget of itself
+	if err := checkSnapshotChangeConflict(st, snapshotID, "forget-snapshot"); err != nil {
+		return nil, nil, err
+	}
+
+	ts := state.NewTaskSet()
+
+	for i, name := range snapNames {
+		desc := fmt.Sprintf("Restoring data from snapshot #%d for snap %q", snapshotID, name)
+		task := st.NewTask("restore-snapshot", desc)
+		shot := shotState{
+			ID:       snapshotID,
+			Snap:     name,
+			Users:    users,
+			Filename: filenames[i],
+		}
+		task.Set("shot", &shot)
+		// hackish, for conflict detection:
+		task.Set("snap-setup", &snapstate.SnapSetup{SideInfo: &snap.SideInfo{RealName: name}})
+		ts.AddTask(task)
+	}
+
+	return snapNames, ts, nil
+}
+
+// Check creates a taskset for checking a snapshot's data.
+// Note that the state must be locked by the caller.
+func Check(st *state.State, snapshotID uint64, snapNames []string, users []string) ([]string, *state.TaskSet, error) {
+	// check needs to conflict with forget of itself
+	if err := checkSnapshotChangeConflict(st, snapshotID, "forget-snapshot"); err != nil {
+		return nil, nil, err
+	}
+
+	snapNames, filenames, err := snapNamesInSnapshot(snapshotID, snapNames)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ts := state.NewTaskSet()
+
+	for i, name := range snapNames {
+		desc := fmt.Sprintf("Checking data in snapshot #%d for snap %q", snapshotID, name)
```

Comment:
"Check data in snapshot set #%d for snap %q".

Reply:
Below you suggest (and I agree!) `"Remove data of snap %q from snapshot set #%d"`, so maybe `"Check data of snap %q from snapshot set #%d"`?

Reply:
(I'll run with it unless you tell me otherwise, given your vacation :-) )